[expo-sms][android] Rewriting expo-sms to kotlin.

kkafar opened the pull request:
# Why

This task was done as a part of migrating expo modules to Kotlin project. 


# How

The code had been initially converted from Java to Kotlin with usage of Android Studio's conversion tool. Then it was manually adjusted / improved. 

# Test Plan

The changes passed 

1. ncl
2. test suite

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).

kkafar committed:
[expo-sms][android] Rewriting expo-sms to kotlin.

kkafar committed:
[expo-sms][android] Removing redundant TODO tag.

kkafar committed:
[expo-sms][android] Rewriting android code to Kotlin -- changelog
update

tsapeta commented on packages/expo-sms/CHANGELOG.md:
```diff
@@ -22,6 +22,8 @@
 
 ### 💡 Others
 
+- Converted Android code to Kotlin ([#13505](https://github.com/expo/expo/pull/13505) by [@kkafar](https://github.com/kkafar))
+
```
There is no need to separate entries with newlines 😉 Also, just so you know — you don't have to add these links manually (this is inconvenient as you have to create the PR first and push once again). You can just add the note and `expo-ci` bot will give you a code review suggestion that you can apply from here

```suggestion
- Converted Android code to Kotlin ([#13505](https://github.com/expo/expo/pull/13505) by [@kkafar](https://github.com/kkafar))
```

tsapeta commented on packages/expo-sms/android/src/main/java/expo/modules/sms/SMSModule.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+    return ""
+  }
+}
```
```suggestion
}

```

tsapeta commented on packages/expo-sms/android/src/main/java/expo/modules/sms/SMSPackage.kt:
```diff
@@ -0,0 +1,11 @@
+  override fun createExportedModules(reactContext: Context): List<ExportedModule> {
+    return listOf(SMSModule(reactContext))
+  }
+}
```
```suggestion
}

```

tsapeta reviewed (approved):
Looks good to me! Congrats on your first contribution 👏

lukmccall reviewed (approved):
Looks good to me 💪  Greate job 🥇

bbarthec commented on packages/expo-sms/android/src/main/java/expo/modules/sms/SMSModule.kt:
```diff
@@ -0,0 +1,134 @@
+package expo.modules.sms
+
+import android.content.Context
+import org.unimodules.core.ExportedModule
+import org.unimodules.core.interfaces.LifecycleEventListener
+import org.unimodules.core.ModuleRegistry
+import org.unimodules.core.Promise
+import org.unimodules.core.interfaces.services.UIManager
+import org.unimodules.core.interfaces.ExpoMethod
+import java.util.ArrayList
+import android.content.Intent
+import android.net.Uri
+import android.provider.Telephony
+import android.content.pm.PackageManager
+import android.os.Bundle
+import org.unimodules.core.interfaces.ActivityProvider
+import java.lang.StringBuilder
```
```suggestion
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.provider.Telephony
import android.os.Bundle

import java.util.ArrayList
import java.lang.StringBuilder

import org.unimodules.core.ExportedModule
import org.unimodules.core.ModuleRegistry
import org.unimodules.core.Promise
import org.unimodules.core.interfaces.ActivityProvider
import org.unimodules.core.interfaces.ExpoMethod
import org.unimodules.core.interfaces.LifecycleEventListener
import org.unimodules.core.interfaces.services.UIManager
```

We need some linter for kotlin files 🤔 @lukmccall, @tsapeta we need to consider adding some in the nearest future 😓

lukmccall replied:
We actually have one, but in Expo Go - https://github.com/expo/expo/blob/30bfe260516b5c4bb3324f8617e561c0e0f50874/android/build.gradle#L120-L124

I'll add it also to the bare-expo

bbarthec commented on packages/expo-sms/android/src/main/java/expo/modules/sms/SMSModule.kt:
```diff
@@ -0,0 +1,134 @@
+
+  override fun onHostDestroy() = Unit
+
+  private fun constructRecipients(addresses: List<String>): String {
+    if (addresses.isNotEmpty()) {
+      val addressesBuilder = StringBuilder(addresses[0])
+      for (address in addresses) {
+        addressesBuilder.append(';').append(address)
+      }
+      return addressesBuilder.toString()
+    }
+    return ""
+  }
```
Can we use https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html to construct this string? 🤔 
cc @lukmccall

lukmccall replied:
Yes, we can, but I'm just curious why we add the first address twice 🤷

bbarthec reviewed (commented):
It looks fine to me, good job 👍 
I'd like you to only research if we can simplify recipients string construction 💪

kkafar committed:
[expo-sms][android] changed SMSModule.constructRecipients method to take
advatage of kotlin .joinToString(...) method && fixed typos && added new
lines at files ends.

kkafar committed:
[expo-sms][android] removed redundant import

bbarthec commented on packages/expo-sms/android/src/main/java/expo/modules/sms/SMSModule.kt:
```diff
@@ -0,0 +1,130 @@
+  override fun onHostDestroy() = Unit
+
+  private fun constructRecipients(addresses: List<String>): String {
+    if (addresses.isNotEmpty()) {
+      return addresses[0] + addresses.joinToString(";")
+    }
+    return ""
```
Is it actually working as expected? 🤔 
I believe not really - [check this playground](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS41LjIwIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiZnVuIG1haW4oKSB7XG4gICAgdmFsIGFkZHJlc3NlcyA9IGxpc3RPZihcIjc4MTI3MzFcIiwgXCIxMjMxMjNcIiwgXCIxMjMxMjFcIilcbiAgICB2YWwgcmVzdWx0ID0gYWRkcmVzc2VzWzBdICsgYWRkcmVzc2VzLmpvaW5Ub1N0cmluZyhcIjtcIilcbiAgICBwcmludGxuKHJlc3VsdClcbn0ifQ==) 😞 

```suggestion
   return addresses.joinToString(";")
```

We can even inline this and get rid of this function altogether 🤔  Let's do so, shall we? 💅

bbarthec commented on packages/expo-sms/android/src/main/java/expo/modules/sms/SMSModule.kt:
```diff
@@ -0,0 +1,130 @@
+    mSMSComposerOpened = false
+  }
+
+  override fun onHostPause() = Unit
+
+  override fun onHostDestroy() = Unit
```
We can safely remove these two, can't we?
```suggestion
```

lukmccall replied:
I believe we can't remove these two functions. The SMS module has to implement the UI listener interface.

bbarthec replied:
😞 Too bad 😞

bbarthec reviewed (changes requested):
There's a problem with constructing recipients list 😞 
Please consider inlining this method as [String#joinToString](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html) is doing all the job properly 💪

kkafar committed:
[sms][android] Construct recipients list with explicit joinToString
calls

private constructRecipients method was removed as redundant -- being only
wrap function for joinToString. I belive this change also improves
readablity.
